### ISSUE TITLE
Increse safety of parameter subsitution?

### ISSUE THREAD

**Issue by BardurArantsson** [lpsmith/postgresql-simple#241]
So, I've been notificing a pattern in some of my own code (and similarly in postgresql-simple code). There seems to be quite a bit of duplication of functions depending on whether or not query substitution is to be performed. For example, in postgresql-simple we have: 

```haskell
query :: (ToRow q, FromRow r) => Connection -> Query -> q -> IO [r]
query_ :: FromRow r => Connection -> Query -> IO [r]

queryWith :: ToRow q => RowParser r -> Connection -> Query -> q -> IO [r]
queryWith_ :: RowParser r -> Connection -> Query -> IO [r]
```

I was just wondering: Do you know if there's any "deep" reason for this? It seems like it would be a straight up win to have separate data constructors, e.g.

```haskell
data Query =
       ParameterizedQuery ByteString (exists p . ToRow p => [p])
    | RawQuery ByteString
```

(I'm not sure the type signature actually works, but the point is that a parameter list only requires a ToRow class. It could perhaps also be split out such that the substitution can be performed separately.)

This is obviously not a huge deal, but it would be nice to avoid the combinatorial "explosion".

Any ideas?

**Comment by lpsmith** [comment]
Yes, you have identified one of my dissatisfactions with the current API, but I haven't thought that much about it.
